internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java

```diff
@@ -90,28 +101,87 @@ public void testBlobP2PTransferAmongServers() throws Exception {
         OffsetRecord offsetServer2 =
-            server2.getVeniceServer().getStorageMetadataService().getLastOffset("test-store_v1", i);
+            server2.getVeniceServer().getStorageMetadataService().getLastOffset("test-store_v1", partitionId);
         Assert.assertEquals(offsetServer1.getLocalVersionTopicOffset(), offsetServer2.getLocalVersionTopicOffset());
       }
     });
   }
 
+  /**
+   * If there are no snapshots available for the store on server2, the blob transfer should throw an exception and return a 404 error.
+   * When server1 restarts and receives the 404 error from server2, it will switch to using Kafka to ingest the data.
+   */
+  @Test(singleThreaded = true)
+  public void testBlobTransferThrowExceptionIfSnapshotNotExisted() throws Exception {
+    String storeName = "test-store-snapshot-not-existed";
+    Consumer<UpdateStoreQueryParams> paramsConsumer = params -> params.setBlobTransferEnabled(true);
+    setUpStore(cluster, storeName, paramsConsumer, properties -> {}, true);
+
+    VeniceServerWrapper server1 = cluster.getVeniceServers().get(0);
+    VeniceServerWrapper server2 = cluster.getVeniceServers().get(1);
+
+    // verify the snapshot is generated for both servers after the job is done
+    for (int i = 0; i < PARTITION_COUNT; i++) {
+      String snapshotPath1 = RocksDBUtils.composeSnapshotDir(path1 + "/rocksdb", storeName + "_v1", i);
+      Assert.assertTrue(Files.exists(Paths.get(snapshotPath1)));
+      String snapshotPath2 = RocksDBUtils.composeSnapshotDir(path2 + "/rocksdb", storeName + "_v1", i);
+      Assert.assertTrue(Files.exists(Paths.get(snapshotPath2)));
+    }
+
+    // cleanup and restart server 1
+    FileUtils.deleteDirectory(
+        new File(RocksDBUtils.composePartitionDbDir(path1 + "/rocksdb", storeName + "_v1", METADATA_PARTITION_ID)));
+
+    List<String> paths = Arrays.asList(path1, path2);
+
+    // clean up all snapshot files on server 1 and server 2,
+    // to simulate the case that snapshot is not existed on server2,
+    // should return 404 in response
+    for (String path: paths) {
+      for (int i = 0; i < PARTITION_COUNT; i++) {
+        FileUtils
+            .deleteDirectory(new File(RocksDBUtils.composePartitionDbDir(path + "/rocksdb", storeName + "_v1", i)));
+        // both partition db and snapshot should be deleted
+        Assert.assertFalse(
+            Files.exists(Paths.get(RocksDBUtils.composePartitionDbDir(path + "/rocksdb", storeName + "_v1", i))));
+        Assert.assertFalse(
+            Files.exists(Paths.get(RocksDBUtils.composeSnapshotDir(path + "/rocksdb", storeName + "_v1", i))));
+      }
+    }
+
+    TestUtils.waitForNonDeterministicAssertion(2, TimeUnit.MINUTES, () -> {
+      cluster.stopAndRestartVeniceServer(server1.getPort());
+      Assert.assertTrue(server1.isRunning());
+    });
+
+    // wait for server 1 to ingest
+    Thread.sleep(120000);
```

Comment:
hmm, this is not favorable. Can we try another way, e.g. check with router or controller to see if server 1 has finished the ingestion?